File: gst-sdk/tutorials/basic-tutorial-13.c
#include <string.h>
#include <gst/gst.h>
  
typedef struct _CustomData {
  GstElement *pipeline;
  GMainLoop *loop;
  
  gboolean playing;  /* Playing or Paused */
  gdouble rate;      /* Current playback rate */
  gboolean backward; /* Forward or backwards */
} CustomData;
  
/* Process keyboard input */
static gboolean handle_keyboard (GIOChannel *source, GIOCondition cond, CustomData *data) {
  gchar *str = NULL;
  
  if (g_io_channel_read_line (source, &str, NULL, NULL, NULL) != G_IO_STATUS_NORMAL) {
    return TRUE;
  }
  
  switch (g_ascii_tolower (str[0])) {
  case 'p':
    data->playing = !data->playing;
    gst_element_set_state (data->pipeline, data->playing ? GST_STATE_PLAYING : GST_STATE_PAUSED);
    g_print ("Setting state to %s\n", data->playing ? "PLAYING" : "PAUSE");
    break;
  case 's':
    if (g_ascii_isupper (str[0])) {
      data->rate *= 2.0;
    } else {
      data->rate /= 2.0;
    }
    gst_element_send_event (data->pipeline,
        gst_event_new_step (GST_FORMAT_TIME, -1, data->rate, TRUE, FALSE));
    g_print ("Current rate: %g\n", data->rate);
    break;
  case 'd':
    data->backward = !data->backward;
    gst_element_send_event (data->pipeline,
        gst_event_new_seek (data->backward ? -data->rate : data->rate,
            GST_FORMAT_TIME, GST_SEEK_FLAG_NONE, GST_SEEK_TYPE_NONE, 0, GST_SEEK_TYPE_NONE, 0));
    g_print ("Going %s\n", data->backward ? "backwards" : "forward");
    break;
  case 'n':
    gst_element_send_event (data->pipeline,
        gst_event_new_step (GST_FORMAT_BUFFERS, 1, data->rate, TRUE, FALSE));
    g_print ("Stepping one frame\n");
    break;
  case 'q':
    g_main_loop_quit (data->loop);
    break;
  default:
    break;
  }
  
  g_free (str);
  
  return TRUE;
}
  
int main(int argc, char *argv[]) {
  CustomData data;
  GstStateChangeReturn ret;
  GIOChannel *io_stdin;
  
  /* Initialize GStreamer */
  gst_init (&argc, &argv);
  
  /* Initialize our data structure */
  memset (&data, 0, sizeof (data));
  
  /* Print usage map */
  g_print (
    "USAGE: Choose one of the following options, then press enter:\n"
    " 'P' to toggle between PAUSE and PLAY\n"
    " 'S' to increase playback speed, 's' to decrease playback speed\n"
    " 'D' to toggle playback direction\n"
    " 'N' to move to next frame (in the current direction, better in PAUSE)\n"
    " 'Q' to quit\n");
  
  /* Build the pipeline */
  data.pipeline = gst_parse_launch ("playbin2 uri=http://docs.gstreamer.com/media/sintel_trailer-480p.webm", NULL);
  
  /* Add a keyboard watch so we get notified of keystrokes */
#ifdef _WIN32
  io_stdin = g_io_channel_win32_new_fd (fileno (stdin));
#else
  io_stdin = g_io_channel_unix_new (fileno (stdin));
#endif
  g_io_add_watch (io_stdin, G_IO_IN, (GIOFunc)handle_keyboard, &data);
  
  /* Start playing */
  ret = gst_element_set_state (data.pipeline, GST_STATE_PLAYING);
  if (ret == GST_STATE_CHANGE_FAILURE) {
    g_printerr ("Unable to set the pipeline to the playing state.\n");
    gst_object_unref (data.pipeline);
    return -1;
  }
  data.playing = TRUE;
  data.rate = 1.0;
  
  /* Create a GLib Main Loop and set it to run */
  data.loop = g_main_loop_new (NULL, FALSE);
  g_main_loop_run (data.loop);
  
  /* Free resources */
  g_main_loop_unref (data.loop);
  g_io_channel_unref (io_stdin);
  gst_element_set_state (data.pipeline, GST_STATE_NULL);
  gst_object_unref (data.pipeline);
  return 0;
}

File: samples/BasicTutorial13.cs
// Authors
//   Copyright (C) 2014 Stephan Sundermann <stephansundermann@gmail.com>

using System;
using Gst;
using System.Runtime.InteropServices;

namespace GstreamerSharp
{
	class Playback
	{
		static bool Playing; // Playing or Paused
		static double Rate; //Current playback rate (can be negative)
		static Element Pipeline, VideoSink;

		// Send seek event to change rate
		static void SendSeekEvent () {
			var format = Format.Time;
			long position;

			// Obtain the current position, needed for the seek event
			if (!Pipeline.QueryPosition (format, out position)) {
				Console.WriteLine ("Unable to retrieve current position.");
				return;
			}

			Event seekEvent;
			// Create the seek event
			if (Rate > 0) {
				seekEvent = Event.NewSeek (Rate, Format.Time, SeekFlags.Flush | SeekFlags.Accurate, SeekType.Set, position, SeekType.None, 0);
			} else {
				seekEvent = Event.NewSeek (Rate, Format.Time, SeekFlags.Flush | SeekFlags.Accurate, SeekType.Set, 0, SeekType.Set, position);
			}

			if (VideoSink == null) {
				// If we have not done so, obtain the sink through which we will send the seek events
				VideoSink = (Element)Pipeline ["video-sink"];
			}

			// Send the event
			VideoSink.SendEvent (seekEvent);

			Console.WriteLine ("Current rate: {0}", Rate);
		}

		// Process keyboard input
		static void HandleKeyboard () {
			ConsoleKeyInfo x;
			bool terminate = false;
			while (!terminate) {
				x = Console.ReadKey ();
				switch (x.Key) {
				case ConsoleKey.P :
					Playing = !Playing;
					Pipeline.SetState (Playing ? State.Playing : State.Paused);
					Console.WriteLine ("Setting state to {0}", Playing ? "PLAYING" : "PAUSE");
					break;
				case ConsoleKey.S:
					if (x.Modifiers == ConsoleModifiers.Shift)
						Rate *= 2.0;
					else
						Rate /= 2.0;
					SendSeekEvent ();
					break;
				case ConsoleKey.D:
					Rate *= -1.0;
					SendSeekEvent ();
					break;
				case ConsoleKey.N:
					if (VideoSink == null) {
						// If we have not done so, obtain the sink through which we will send the step events
						VideoSink = (Element)Pipeline ["video-sink"];
					}
					var evnt = Event.NewStep (Format.Buffers, 1, Rate, true, false);
					VideoSink.SendEvent (evnt);

					Console.WriteLine ("Stepping one frame");
					break;
				case ConsoleKey.Q:
					terminate = true;
					break;
				default:
					break;
				}
			}
		}

		public static void Main (string[] args)
		{
			// Initialize GStreamer
			Application.Init (ref args);

			// Print usage map
			Console.WriteLine ("USAGE: Choose one of the following options, then press enter:");
			Console.WriteLine (" 'P' to toggle between PAUSE and PLAY");
			Console.WriteLine (" 'S' to increase playback speed, 's' to decrease playback speed");
			Console.WriteLine (" 'D' to toggle playback direction");
			Console.WriteLine (" 'N' to move to next frame (in the current direction, better in PAUSE)");
			Console.WriteLine (" 'Q' to quit");

			// Build the pipeline
			//Pipeline = Parse.Launch ("playbin uri=http://download.blender.org/durian/trailer/sintel_trailer-1080p.mp4");
			Pipeline = Parse.Launch ("playbin uri=file:///home/stephan/Downloads/sintel_trailer-1080p.mp4");

			// Start playing
			var ret = Pipeline.SetState (State.Playing);
			if (ret == StateChangeReturn.Failure) {
				Console.WriteLine ("Unable to set the pipeline to the playing state.");
				return;
			}
			Playing = true;
			Rate = 1.0;

			// Process input
			HandleKeyboard ();

			// Free resources
			Pipeline.SetState (State.Null);
		}
	}
}